modules/fedimint-ln-client/src/recurring/mod.rs

```diff
@@ -0,0 +1,525 @@
+    }
+
+    pub async fn subscribe_ln_recurring_receive(
+        &self,
+        operation_id: OperationId,
+    ) -> anyhow::Result<UpdateStreamOrOutcome<LnReceiveState>> {
+        let operation = self.client_ctx.get_operation(operation_id).await?;
+        let LightningOperationMetaVariant::RecurringPaymentReceive(ReurringPaymentReceiveMeta {
+            invoice,
+            ..
+        }) = operation.meta::<LightningOperationMeta>().variant
+        else {
+            bail!("Operation is not a recurring lightning receive")
+        };
+
+        let client_ctx = self.client_ctx.clone();
+
+        Ok(self.client_ctx.outcome_or_updates(operation, operation_id, move || {
+            stream! {
+                let self_ref = client_ctx.self_ref();
+
+                yield LnReceiveState::Created;
+                yield LnReceiveState::WaitingForPayment { invoice: invoice.to_string(), timeout: invoice.expiry_time() };
+
+                match self_ref.await_receive_success(operation_id).await {
+                    Ok(_) => {
+                        yield LnReceiveState::Funded;
+
+                        if let Ok(out_points) = self_ref.await_claim_acceptance(operation_id).await {
+                            yield LnReceiveState::AwaitingFunds;
+
+                            if client_ctx.await_primary_module_outputs(operation_id, out_points).await.is_ok() {
+                                yield LnReceiveState::Claimed;
+                                return;
+                            }
+                        }
+
+                        yield LnReceiveState::Canceled { reason: LightningReceiveError::Rejected };
+                    }
+                    Err(e) => {
+                        yield LnReceiveState::Canceled { reason: e };
+                    }
+                }
+            }
+        }))
+    }
+
+    pub async fn list_recurring_payment_codes(&self) -> BTreeMap<u64, RecurringPaymentCodeEntry> {
+        self.client_ctx
+            .module_db()
+            .begin_transaction_nc()
+            .await
+            .find_by_prefix(&RecurringPaymentCodeKeyPrefix)
+            .await
+            .map(|(idx, entry)| (idx.derivation_idx, entry))
+            .collect()
+            .await
+    }
+
+    pub async fn get_recurring_payment_code(
+        &self,
+        payment_code_idx: u64,
+    ) -> Option<RecurringPaymentCodeEntry> {
+        self.client_ctx
+            .module_db()
+            .begin_transaction_nc()
+            .await
+            .get_value(&RecurringPaymentCodeKey {
+                derivation_idx: payment_code_idx,
+            })
+            .await
+    }
+
+    pub async fn list_recurring_payment_code_invoices(
+        &self,
+        payment_code_idx: u64,
+    ) -> Option<BTreeMap<u64, OperationId>> {
+        let payment_code = self.get_recurring_payment_code(payment_code_idx).await?;
+
+        let operations = (1..=payment_code.last_derivation_index)
```

Comment:
Why does this start at 1?

Reply:
0 is a special value indicating no invoices were generated yet, not great I guess, but also not sure if worth fixing …